synapse/storage/databases/state/store.py

```diff
@@ -279,7 +300,12 @@ def _get_state_for_group_gather_inflight_requests(
 
         # The list of ongoing requests which will help narrow the current request.
         reusable_requests = []
-        for (request_state_filter, request_deferred) in inflight_requests.items():
+
+        # Iterate over existing requests in roughly biggest-first order.
+        # reversed(inflight_requests) has an efficient iterator implementation,
+        # but reversed(inflight_requests.items()) does not, sadly.
```

Comment:
Not sure if it helps, but you could negate the values of `state_filter_rough_size_comparator` (or return 0 for `include_other` / 1 otherwise; `-len(state_filter.types)`) to "automatically" reverse this and not need to worry about the perf of `reversed`.

Reply:
Yeah, maybe that will be better. Let's give it a shot.